### ISSUE TITLE
feature request: Support JUnit5 @BeforeAll, @BeforeEach, @AfterEach and @AfterAll callbacks

### ISSUE THREAD

**Issue by beskow** [database-rider/database-rider#207]
Minimizing redundancy in test data setup is important for readability and maintainability, which is the reason behind JUnit 5's lifecycle mechanism. There is however only limited support for the JUnit5 callbacks in current version of DBRider. The @BeforeEach callback can be annotated with a @DataSet annotation, but any @DBUnit annotation on that method or class is ignored (and hence only config via global dbunit.yml file is possible). The other lifecycle callbacks (@BeforeAll, @AfterEach and @AfterAll) are not supported.

To be fully functional, all lifecycle callbacks should be supported in the same way as regular test methods are, as in the example below:

~~~
@DBRider
@DBUnit(url = "jdbc:hsqldb:mem:junit5;DB_CLOSE_DELAY=-1", driver = "org.hsqldb.jdbcDriver", user = "sa")
@SpringBootTest
public class SpringBootDBUnitTest {

    @Autowired
    private UserRepository userRepository;

    @BeforeAll
    @DataSet("emptyUsers.yml")
    public static void setUp() {
    }

    @AfterAll
    @DataSet("emptyUsers.yml")
    public static void cleanUp() {
    }

    @BeforeEach
    @DataSet("users.yml")
    public void setUpUsers() {
    }

    @AfterEach
    @ExpectedDataSet("expectedUsers.yml")
    public void verifyInvariant() {
    }

    @Test
    public void shouldListUsers() {
        assertThat(userRepository).isNotNull();
        assertThat(userRepository.count()).isEqualTo(3);
        assertThat(userRepository.findByEmail("springboot@gmail.com")).isEqualTo(new User(3));
    }
   ...
}
~~~

**Comment by rmpestano** [comment]
@beskow, 

I think seeding DB after the test may cause more harm than good because next test to execute may not be expecting the data you seed in after each. So it is preferable to prepare all db state you need before the test execution. 

Do you have a specific use-case, besides the example you posted?  

Thank you for the feedback!

**Comment by beskow** [comment]
Hi @rmpestano, thanks for the quick response!

In the example I gave (from the example project), I just tried to examplify the syntax.

Just as you point out, the most frequently used seeding callback is `@BeforeEach`. In some of our tests that are on the borderline between functional and performance tests, we use a large set of background data which in essence is read-only. Seeding this data for each test is obviously safer (if some test by mistake changes the data), but leads to very long execution times. Hence I think it makes perfect sense to be able to seed that data once for a test, using `@BeforeAll`. Also for completeness' sake: it puzzled me that a `@DataSet` annotation on a beforeAll method had no effect, since the `@BeforeAll` and `@AfterAll` methods are part of the standard JUnit API. Don't you agree?

Seeding data after a test may look awkward, I agree. We use it to ensure a clean state after test execution (I.e. we typically only use data sets with empty tables to effectively clean the database content). It can be done with an SQL script instead, but is convenient to be able to do via a dataset.

Regarding `@ExpectedDataSet`, I think it also makes sense to be able to specify "invariants" for data sets: "Regardless of which test scenario, this data should always be present". We don't use such invariants very frequently, but when we do, it is quite handy since it eliminates redundancy and highlights the invariant.

So the rationale behind the suggestion I guess is the same rationale used to motivate the standard lifecycle callback methods in the JUnit 5 API. Since they are there, they should be useable with DBRider as well, I think.

It was straight-forward to implement, so I have a patch ready which adds the requested feature. I'll submit it as a pull request. Let me know what you think!

Again, thanks for an excellent framework! It brings a lot of value to my current project.

Cheers
/Björn
